docs/RFCS/20171101_encryption_at_rest.md

```diff
@@ -0,0 +1,1098 @@
+monotonically increasing number. Upon successful write, we trigger a data key file reload.
+
+The actual process of creating the new file should match the rocksdb method (write to temporary file and rename).
+We need to guarantee that the latest file is correct.
+
+Key generation is done inline at startup (we may as well wait for the new key before proceeding), but in the
+background for automated changes while the system is already running.
+
+### Reporting encryption status
+
+We need to report basic information about the current status of encryption.
+
+At the very least, we should have:
+* log entries
+* debug page entries per store
+
+With the following information:
+* preamble format support
+* active store key ID and cipher
+* active data key ID and cipher
+* fraction of live data per key ID and cipher
+
+We can report the following encryption status:
+* `PLAIN`: plaintext data
+* `AES-<size>`: encrypted with AES (one entry for each key size)
+* `AES-<size> EXPOSED`: encrypted, but data key was exposed at some point
+
+Preamble format support and active key IDs and ciphers are known at all times. We need to log them when they change
+(indicating successful key rotation) and propagate the information to the Go layer.
+
+Fraction of data encoded is a bit trickier. We need to:
+1. find all files in use
+1. read file preambles for encryption status (key ID and cipher)
+1. determine file sizes (without the preamble)
+1. log a summary
+1. report back to the go layer
+
+We can find the list of all in-use files the same way rocksdb's backup does, by calling:
+* `rocksdb::GetLiveFiles`: retrieve the list of all files in the database
+* `rocksdb::GetSortedWalFiles`: retrieve the sorted list of all wal files
+
+Reading each file preamble means reading the first 4KiB of each file. To avoid performing this too
+frequently, we can cache results. We must take particular care with files being overwritten (eg: `CURRENT`) and
+using a new key/cipher.
+
+### Other uses of local disk
+
+**Note: logs encryption is currently [Out of scope](#out-of-scope)**
+
+All existing uses of local disk to process data must apply the desired encryption status.
+
+Data tied to a specific store should use the store's rocksdb instance for encryption.
+Data not necessarily tied to a store should be encrypted if any of the stores on the node is encrypted.
+
+We identify some existing uses of local disk:
+TODO(mberhault, mjibson, dan): make sure we don't miss anything.
+
+1. temporary work space for dist SQL: written through a temporary instance of rocksdb. This data does not need
+to be used by another rocksdb instance and does not survive node restart. We propose to use dynamically-generated
+keys to encrypt the temporary rocksdb instance.
+1. sideloading for restore. Local SSTables are generated using an in-memory rocksdb instance then written in go
+to local disk. We must change this to either be written directly by rocksdb, or move encryption to Go. The former
```

Comment:
Not sure what exactly you mean, not putting the sideloaded file on disk directly is likely a non-started, since avoiding the write amplification associated with putting it into RocksDB was one of the goals of side loading.

Also, how will RESTORE work in this case? The sideloaded files are linked in directly into RocksDB (via `DB::IngestExternalFile()`). What does that mean for the file? Does it have to be encrypted already?

My assumption on how this would work is that instead of slapping the file on disk, it would pass through something that would add the right preamble and encrypt with the store key, and that that is the right thing to pass to `IngestExternalFile` later, but I'm not sure.

Could you add more details here?

Reply:
As mentioned in the RFC, the file is generated in an in-mem rocksdb instance then written to disk in Go. After that, it is ingested.
We need to make sure the ingested file is encrypted using the same set of keys, so it really needs to go through the existing rocksdb instance, or at least the same `Env` layer.